### PR TITLE
Add Couchbase cache plugin

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
- based on Couchbase Java SDK 2.7.4

### CODE REVIEW

File: parent/pom.xml [hunk 1]
```diff
@@ -85,6 +85,8 @@
    <properties>
       <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
       <plugins.chm/>
+      <plugins.couchbase/>
+      <plugins.couchbase-latest/>
```
Comment (daschl) [inline parent/pom.xml]: why is this needed?

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseService.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+   @Property(doc = "Password")
+   protected String password = "password";
+
+   @Property(doc = "Bucket name")
+   protected String bucketName = "default";
+
+   @Override
+   public void start() {
+      cluster = CouchbaseCluster.create(nodes);
```
Comment (daschl) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseService.java]: I think since this is just doing sync kv ops, we can pass in the environment and set the following:

.callbacksOnIoPool(true)
.runtimeMetricsCollectorConfig(DefaultMetricsCollectorConfig.disabled())
.networkLatencyMetricsCollectorConfig(DefaultLatencyMetricsCollectorConfig.disabled())

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      @Override
+      public byte[] getAndRemove(byte[] key) {
+         byte[] doc = get(key);
+         remove(key);
+         return doc;
+      }
+
+      @Override
+      public void clear() {
+         throw new UnsupportedOperationException("Clearing Couchbase is not implemented yet.");
```
Comment (daschl) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: does this compare to the flush() command on the bucket manager?

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+      @Override
+      public void put(byte[] key, byte[] value) {
+         ByteArrayDocument doc = ByteArrayDocument.create(new String(key), value);
+         service.bucket.upsert(doc);
+      }
+
+      @Override
+      public byte[] getAndPut(byte[] key, byte[] value) {
+         byte[] document = get(key);
```
Comment (daschl) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: I think this should use the cas on put (replace) to be more realistic.

Might make sense to have specialized impls for this to avoid creating a string key more than once

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            service.bucket.remove(new String(key));
+            return true;
+         } catch(Exception e) {
+            return false;
+         }
+      }
+
+      @Override
+      public byte[] getAndRemove(byte[] key) {
+         byte[] doc = get(key);
```
Comment (daschl) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: I think this should use the cas on remove to be more realistic.

Might make sense to have specialized impls for this to avoid creating a string key more than once

File: parent/pom.xml (hunk 1 above)
Comment (MikeGoldsmith) [reply parent/pom.xml]: I believe this is used to discover the plugin projects and is used by the build scripts when creating the distributable ZIP.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java (hunk 3 above)
Comment (MikeGoldsmith) [reply plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: Yes, it does - I didn't implement as the Redis plugin provider I used as a basis didn't implement it. I'm not sure if it's used but can update to do the flush.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java (hunk 4 above)
Comment (MikeGoldsmith) [reply plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: Sure, will look to do that.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java (hunk 5 above)
Comment (MikeGoldsmith) [reply plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: Sure, will look to do that.

File: parent/pom.xml (hunk 1 above)
Comment (alanfx) [reply parent/pom.xml]: First of all, thank you for this contribution. This property allows you to build a RadarGun distribution that only includes the latest version of each plugin, instead of every version of the plugin.

**Comment by MikeGoldsmith** [comment]
@alanfx  @diegolovison I think this is ready for initial review. Let me know if there is anything else you would like to see in this PR.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+   public CouchbaseOperations(CouchbaseService service) {
+      this.service = service;
+   }
+
+   @Override
+   public <K, V> Cache<K, V> getCache(String cacheName) {
+      return adapter;
+   }
+
+   private class CouchbaseCacheAdapter<K, V> implements BasicOperations.Cache<byte[], byte[]> {
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: You should use something like: `private class CouchbaseCacheAdapter <K, V> implements BasicOperations.Cache<K, V>`. See as an example: https://github.com/radargun/radargun/commit/373b353317d0a05b9a7e9e7009a3c376500b8d24#diff-df93a8f6e3d2569df9283e049b55bfbc

RadarGun allows different keys and values generators. Here are some that we support https://github.com/radargun/radargun/tree/master/extensions/cache/src/main/java/org/radargun/stages/cache/generators

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
+package org.radargun.service;
+
+import com.couchbase.client.java.document.ByteArrayDocument;
+import org.radargun.traits.BasicOperations;
+
+import java.nio.charset.StandardCharsets;
+
+public class CouchbaseOperations implements BasicOperations {
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: I prefer to see the name of the class like `CouchbaseClientOperations`

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseService.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.CouchbaseCluster;
+import com.couchbase.client.java.env.CouchbaseEnvironment;
+import com.couchbase.client.java.env.DefaultCouchbaseEnvironment;
+import org.radargun.Service;
+import org.radargun.config.Property;
+import org.radargun.traits.Lifecycle;
+
+@Service(doc = "Couchbase")
+public class CouchbaseService implements Lifecycle {
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseService.java]: Same here: CouchbaseClientService

File: plugins/couchbase/src/main/resources/plugin.properties [hunk 9]
```diff
@@ -0,0 +1,3 @@
+service.default org.radargun.service.CouchbaseService
```
Comment (diegolovison) [inline plugins/couchbase/src/main/resources/plugin.properties]: For me, it is missing the `CouchbaseServerService` class implementation. For example, I would like to run RadarGun and compare the project X with Couchbase in a perf lab. Without `CouchbaseServerService` I will need to setup the server everytime. I saw https://docs.couchbase.com/server/4.1/getting-started/installing.html that we can download and install the server. A good entry point is https://github.com/radargun/radargun/blob/master/plugins/etcd/src/main/java/org/radargun/service/EtcdServerService.java

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      private void put(String key, byte[] value) {
+         ByteArrayDocument doc = ByteArrayDocument.create(key, value);
+         service.bucket.upsert(doc);
+      }
+
+      @Override
+      public byte[] getAndPut(byte[] keyBytes, byte[] value) {
+         String key = createKey(keyBytes);
+         byte[] document = get(key);
+         put(key, value);
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java]: @rvansa should it be in a unique operation or doesn't support it, right?

**Comment by MikeGoldsmith** [comment]
@diegolovison I've updated the class names / files and refactored `CouchbaseClientOperations` to utilize **K** and **V** type params in the inner `CouchbaeCacheAdapter`.

I'm not sure downloading and installing Couchbase as part of a test run is a good idea. It is a fairly large process / download. There are docker images and/or vagrant scripts that we may be able to utilize.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+            return true;
+         } catch(Exception e) {
+            return false;
+         }
+      }
+
+      @Override
+      public V getAndRemove(K key) {
+         String keyStr = keyAsString(key);
+         V document = get(keyStr);
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java]: It should be only one operation in the server. In this case, we are calling the server twice. If Couchbase doesn't support returning the previous value when removing the key we need to throw an exception.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+      private void put(String key, V value) {
+         ByteArrayDocument doc = ByteArrayDocument.create(key, valueAsByteArray(value));
+         service.bucket.upsert(doc);
+      }
+
+      @Override
+      public V getAndPut(K key, V value) {
+         String keyStr = keyAsString(key);
+         V document = get(keyStr);
+         put(keyStr, value);
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java]: It should be only one operation in the server. In this case, we are calling the server twice. If Couchbase doesn't support returning the previous value when putting the key we need to throw an exception.

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
+package org.radargun.service;
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java]: I just saw that the folder structure is using (dot) instead of a folder separator.
The name of the file is: `org.radargun.service/CouchbaseClientOperations.java` and it should be `org/radargun/service/CouchbaseClientOperations.java`

File: plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientService.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
+package org.radargun.service;
```
Comment (diegolovison) [inline plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientService.java]: Same here about the folder separator

**Comment by diegolovison** [comment]
@MikeGoldsmith I am going to squash and merge using the GitHub button. IIRC, the commit will be in my name. I would like to see the commit with who developed that. Would you like to squash or can I move forward?

**Comment by diegolovison** [comment]
@MikeGoldsmith I was going to use the "Squash and merge" button in the GitHub interface and I just remembered that last time that I did it, I was the commit author. I prefer to see your team as an author of this commit. Could you please squash?

**Comment by MikeGoldsmith** [comment]
@diegolovison commits have been squashed

**Comment by MikeGoldsmith** [comment]
@diegolovison commits have been squashed 👍

**Comment by diegolovison** [comment]
Thanks @MikeGoldsmith and @daschl
